packages/salesforcedx-vscode-core/src/commands/commands.ts

```diff
@@ -0,0 +1,147 @@
+//////////////////
+export interface ContinueResponse<T> {
+  type: 'CONTINUE';
+  data: T;
+}
+
+export interface CancelResponse {
+  type: 'CANCEL';
+}
+
+export interface ParametersGatherer<T> {
+  gather(): Promise<CancelResponse | ContinueResponse<T>>;
+}
+
+export class EmptyParametersGatherer implements ParametersGatherer<{}> {
+  public async gather(): Promise<CancelResponse | ContinueResponse<{}>> {
+    return { type: 'CONTINUE', data: {} };
+  }
+}
+
+export type FileSelection = { file: string };
+export class FileSelector implements ParametersGatherer<FileSelection> {
+  private readonly include: string;
+  private readonly exclude?: string;
+  private readonly maxResults?: number;
+
+  constructor(include: string, exclude?: string, maxResults?: number) {
+    this.include = include;
+    this.exclude = exclude;
+    this.maxResults = maxResults;
+  }
+
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<FileSelection>
+  > {
+    const files = await vscode.workspace.findFiles(
+      this.include,
+      this.exclude,
+      this.maxResults
+    );
+    const fileItems = files.map(file => {
+      return {
+        label: path.basename(file.toString()),
+        description: file.fsPath
+      };
+    });
+    const selection = await vscode.window.showQuickPick(fileItems);
+    return selection
+      ? { type: 'CONTINUE', data: { file: selection.description.toString() } }
+      : { type: 'CANCEL' };
+  }
+}
+
+// Command Execution
+////////////////////
+export interface CommandletExecutor<T> {
+  execute(response: ContinueResponse<T>): void;
+}
+
+// Common
+
+export abstract class SfdxCommandletExecutor<T>
+  implements CommandletExecutor<T> {
+  public execute(response: ContinueResponse<T>): void {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
+    }).execute(cancellationToken);
+
+    channelService.streamCommandOutput(execution);
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+    CancellableStatusBar.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  public abstract build(data: T): Command;
+}
+
+export class SfdxCommandlet<T> {
+  private readonly checker: PreconditionChecker;
+  private readonly gatherer: ParametersGatherer<T>;
+  private readonly executor: CommandletExecutor<T>;
+
+  constructor(
+    checker: PreconditionChecker,
+    gatherer: ParametersGatherer<T>,
+    executor: CommandletExecutor<T>
+  ) {
+    this.checker = checker;
+    this.gatherer = gatherer;
+    this.executor = executor;
+  }
+
+  public async run(): Promise<void> {
+    if (this.checker.check()) {
+      const inputs = await this.gatherer.gather();
+
+      switch (inputs.type) {
```

Comment:
Here is where we check for the type from the input gatherer.
TypeScript doesn't have pattern matching but this switch case is meant to simulate that.
This does _exhaustive_ checking over the data type that we expect.

If it is not a cancellation then we can proceed safely (and know that data will not be null).
If it is cancellation then we just don't proceed.